src/mlpack/methods/ann/image_functions/bilinear_function.hpp

```diff
@@ -0,0 +1,170 @@
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_IMAGE_FUNCTIONS_BILINEAR_FUNCTION_HPP
+#define MLPACK_METHODS_ANN_IMAGE_FUNCTIONS_BILINEAR_FUNCTION_HPP
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * The bilinear interpolation function
+ *
+ */
+class BiLinearFunction
+{
+ /*
+  * The constructor for the bilinear interpolation.
+  * 
+  * @param inRowSize: Number of input rows
+  * @param inColSize: Number of input columns.
+  * @param outRowSize: Number of output rows.
+  * @param outColSie: Number of output columns.
+  */
+ public:
+  BiLinearFunction(
+      const size_t inRowSize,
+      const size_t inColSize,
+      const size_t outRowSize,
+      const size_t outColSize):
+      inRowSize(inRowSize),
+      inColSize(inColSize),
+      outRowSize(outRowSize),
+      outColSize(outColSize)
+  {};
+  /**
+   * UpSample the given input.
+   *
+   * @param input The input matrix
+   * @param output The resulting interpolated output image.
+   */
+  template<typename eT>
+  void UpSample(const arma::Mat<eT>& input, arma::Mat<eT>& output)
+  {
+    // Get dimensions.
+    if (output.is_empty() && (output.n_cols != outColSize || output.n_rows != outRowSize))
+      output = arma::resize(output, outRowSize, outColSize);
+
+    scaleRow = (double)input.n_rows / (double) output.n_rows;
+    scaleCol = (double)input.n_cols / (double) output.n_cols;
+
+    for (size_t i = 0; i < output.n_rows; i++)
+    {
+      for (size_t j = 0; j < output.n_cols; j++)
+      {
+        rOrigin = std::floor(i * scaleRow);
+        cOrigin = std::floor(j * scaleCol);
+
+        if (rOrigin > input.n_rows - 2)
+          rOrigin = input.n_rows - 2;
+        if (cOrigin > input.n_cols - 2)
+          cOrigin = input.n_cols - 2;
+        
+
+        double deltaR = i * scaleRow - rOrigin;
+        double deltaC = j * scaleCol - cOrigin;
+        coeff1 = (1 - deltaR) * (1 - deltaC);
+        coeff2 = deltaR * (1 - deltaC);
+        coeff3 = (1 - deltaR) * deltaC;
+        coeff4 = deltaR * deltaC;
+
+
+        output(i, j) =  input(cOrigin * input.n_rows + rOrigin) * coeff1 +
+                        input(cOrigin * input.n_rows + rOrigin + 1) * coeff2 +
+                        input((cOrigin + 1) * input.n_rows + rOrigin) * coeff3 +
+                        input((cOrigin + 1) * input.n_rows + rOrigin+1) * coeff4;
+      }
+    }
+  }
+
+  /**
+   * DownSample the given input.
+   *
+   * @param input The input matrix
+   * @param output The resulting down-sampled output image.
+   */
+  template<typename eT>
+  void DownSample(const arma::Mat<eT>& input, arma::Mat<eT>& output)
+  {
+
+    if (output.is_empty() && (output.n_cols != inColSize || output.n_rows != inRowSize))
+      output = arma::resize(output, inRowSize, inColSize);
```

Comment:
I am not sure that the layer should reshape the output. Maybe it is better to represent the output as a column `(inRowSize * inColSize, 1)`. In that case the layer should work with the linear layer.